<!DOCTYPE html>

<html lang="en-us">
    <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="format-detection" content="telephone=no"/>

    <title> | Lexi&#39;s zone</title>
    
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
    <link rel="manifest" href="/manifest.json">
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#FF3DB4">
    <meta name="theme-color" content="#ffffff">

    
    
    
    <link rel="stylesheet" href="https://lexinwei.github.io/css/main.min.7f48fb61d97a236c01a3d2327f63d7bc0f1b5da94c7eb37c0e7b8f0a7847b715.css"/>

    
    
    

    
    
 
    </head>

    <body>
        
<nav>
  <header>
    <div class="site-title">
        <a href="/">Lexi&#39;s zone</a>
    </div>  
</header>

  <div class="nav-menu">
  
    <a class="color-link nav-link" href="/about/">About</a>
  
    <a class="color-link nav-link" href="/tags/">Tags</a>
  
    <a class="color-link nav-link" href="/archives/">Archives</a>
  
    <a class="color-link nav-link" href="/about/">About</a>
  
  <a class="color-link nav-link" href="https://lexinwei.github.io/index.xml" target="_blank" rel="noopener" type="application/rss+xml">RSS</a>
</div>
<footer class="footer">
	<div class="social-icons">
        
    <a class="social-icon" href="mailto:xwei97@zju.edu.cn" target="_blank" rel="noopener" title="Email">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M25.2794292,5.59128519 L14,16.8707144 L2.72057081,5.59128519 C3.06733103,5.30237414 3.51336915,5.12857603 4,5.12857603 L24,5.12857603 C24.4866308,5.12857603 24.932669,5.30237414 25.2794292,5.59128519 Z M25.9956978,6.99633695 C25.998551,7.04004843 26,7.08414302 26,7.12857603 L26,20.871424 C26,21.0798433 25.9681197,21.2808166 25.9089697,21.4697335 L18.7156355,14.2763993 L25.9956978,6.99633695 Z M24.9498374,22.6319215 C24.6672737,22.7846939 24.3437653,22.871424 24,22.871424 L4,22.871424 C3.5268522,22.871424 3.09207889,22.7071233 2.74962118,22.432463 L10.0950247,15.0870594 L13.9848068,18.9768415 L14.1878486,18.7737996 L14.2030419,18.7889929 L17.6549753,15.3370594 L24.9498374,22.6319215 Z M2.00810114,21.0526627 C2.00273908,20.9929669 2,20.9325153 2,20.871424 L2,7.12857603 C2,7.08414302 2.00144896,7.04004843 2.00430222,6.99633695 L9.03436454,14.0263993 L2.00810114,21.0526627 Z"></path>
        </svg>
    </a>
    

    

    
    <a class="social-icon" href="https://twitter.com/LexiWeiXin" target="_blank" rel="noopener" title="Twitter">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M8.991284,24.971612 C19.180436,24.971612 24.752372,16.530224 24.752372,9.210524 C24.752372,8.970656 24.747512,8.731868 24.736496,8.494376 C25.818008,7.712564 26.758256,6.737 27.5,5.62622 C26.507372,6.067076 25.439252,6.364292 24.318752,6.498212 C25.462472,5.812628 26.340512,4.727444 26.754584,3.434036 C25.684088,4.068536 24.499004,4.53002 23.23724,4.778528 C22.226468,3.701876 20.786828,3.028388 19.193828,3.028388 C16.134404,3.028388 13.653536,5.509256 13.653536,8.567492 C13.653536,9.0023 13.702244,9.424904 13.797176,9.830552 C9.19346,9.599108 5.11106,7.39472 2.3792,4.04294 C1.903028,4.861364 1.629032,5.812628 1.629032,6.827072 C1.629032,8.74904 2.606972,10.445612 4.094024,11.438132 C3.185528,11.41016 2.331788,11.160464 1.585184,10.745096 C1.583888,10.768208 1.583888,10.791428 1.583888,10.815728 C1.583888,13.49888 3.493652,15.738584 6.028088,16.246508 C5.562932,16.373084 5.07326,16.44134 4.56782,16.44134 C4.210988,16.44134 3.863876,16.406024 3.526484,16.34144 C4.231724,18.542264 6.276596,20.143796 8.701412,20.18894 C6.805148,21.674696 4.416836,22.56008 1.821488,22.56008 C1.374476,22.56008 0.93362,22.534592 0.5,22.4834 C2.951708,24.054476 5.862524,24.971612 8.991284,24.971612"></path>
        </svg>
    </a>
    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
    
    
    <a class="social-icon" href="https://github.com/lexinwei/" target="_blank" rel="noopener" title="GitHub">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M13.9988029,1.32087331 C6.82105037,1.32087331 1,7.14112562 1,14.3212723 C1,20.0649109 4.72454649,24.9370678 9.89038951,26.6560892 C10.5408085,26.7757983 10.7778323,26.374374 10.7778323,26.0296121 C10.7778323,25.7215609 10.7666595,24.9035493 10.760275,23.8189856 C7.14426471,24.6042767 6.38131925,22.0760223 6.38131925,22.0760223 C5.78995672,20.5740732 4.93762853,20.1742451 4.93762853,20.1742451 C3.75729765,19.3682044 5.02701126,19.3841656 5.02701126,19.3841656 C6.33183953,19.4759425 7.01817121,20.7241085 7.01817121,20.7241085 C8.17775254,22.7104801 10.0611744,22.1366749 10.8017741,21.8038838 C10.919887,20.9643246 11.2558703,20.3913175 11.6269683,20.066507 C8.74038491,19.7385043 5.70536235,18.6228163 5.70536235,13.6413251 C5.70536235,12.2223743 6.21213051,11.0611968 7.04370914,10.1530044 C6.90963504,9.82420367 6.46351945,8.50181809 7.17139875,6.71256734 C7.17139875,6.71256734 8.26234691,6.36301702 10.7459099,8.04532771 C11.78259,7.75642995 12.8950858,7.61277914 14.000399,7.60719272 C15.1049142,7.61277914 16.2166119,7.75642995 17.2548881,8.04532771 C19.736855,6.36301702 20.8262071,6.71256734 20.8262071,6.71256734 C21.5356825,8.50181809 21.0895669,9.82420367 20.9562909,10.1530044 C21.7894656,11.0611968 22.2922435,12.2223743 22.2922435,13.6413251 C22.2922435,18.6355852 19.2524325,19.734514 16.3570705,20.0561322 C16.8231376,20.4575564 17.2389269,21.2508282 17.2389269,22.4638795 C17.2389269,24.2012564 17.2229657,25.603448 17.2229657,26.0296121 C17.2229657,26.3775663 17.4575954,26.7821827 18.116793,26.6552912 C23.2786458,24.9322794 27,20.0633148 27,14.3212723 C27,7.14112562 21.1789496,1.32087331 13.9988029,1.32087331"></path>
        </svg>
    </a>
    

    
    <a class="social-icon" href="https://stackoverflow.com/users/11497074/x-wei" target="_blank" rel="noopener" title="Stack Overflow">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M20.8863636,23.7090909 L20.8863636,17.3818182 L22.9863636,17.3818182 L22.9863636,25.8090909 L4.03181818,25.8090909 L4.03181818,17.3818182 L6.13181818,17.3818182 L6.13181818,23.7090909 L20.8863636,23.7090909 Z M8.45,16.7818182 L8.88636364,14.7090909 L19.1954545,16.8636364 L18.7590909,18.9363636 L8.45,16.7818182 Z M9.81363636,11.8727273 L10.6863636,9.93636364 L20.2318182,14.4090909 L19.3590909,16.3181818 L9.81363636,11.8727273 Z M12.4590909,7.18181818 L13.7954545,5.57272727 L21.8954545,12.3090909 L20.5590909,13.9181818 L12.4590909,7.18181818 Z M17.6954545,2.19090909 L23.9681818,10.6454545 L22.2772727,11.9 L16.0045455,3.44545455 L17.6954545,2.19090909 Z M8.23181818,21.5818182 L8.23181818,19.4818182 L18.7590909,19.4818182 L18.7590909,21.5818182 L8.23181818,21.5818182 Z"></path>
        </svg>
    </a>
    
    
    

    

    
    <a class="social-icon" href="https://scholar.google.com/citations?user=r7VSdzkAAAAJ&amp;hl=en" target="_blank" rel="noopener" title="Google Scholar">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
             <path d="M15.7399177,5.48151786 C15.149924,4.82272859 14.4663239,4.49480724 13.6933878,4.49480724 C12.8185607,4.49480724 12.1471312,4.80966127 11.6791845,5.43796013 C11.2112379,6.06433732 10.9771365,6.82326666 10.9771365,7.71483354 C10.9771365,8.47423262 11.1052474,9.24733956 11.3616402,10.035649 C11.6170934,10.8236168 12.0350768,11.5270314 12.6210136,12.1470458 C13.2049434,12.7685974 13.8844439,13.0787968 14.6564833,13.0787968 C15.5171327,13.0787968 16.189459,12.7903763 16.671626,12.2137916 C17.1517432,11.6379755 17.3927199,10.9111166 17.3927199,10.035649 C17.3927199,9.29000051 17.2655484,8.50924961 17.0102233,7.69296927 C16.7561365,6.87562134 16.3319184,6.13812924 15.739875,5.48156056 L15.7399177,5.48151786 Z M24.3072528,4.96950105 L24.3072528,13.3462071 C24.3072528,13.7339989 23.9898792,14.0515006 23.6019592,14.0515006 L23.3422783,14.0515006 C22.9543156,14.0515006 22.6369848,13.7341271 22.6369848,13.3462071 L22.6369848,4.96950105 C22.6369848,4.62513879 22.5869788,4.33765779 23.068249,4.27714671 L23.068249,3.16262405 L19.3634081,6.20120253 C19.4061972,6.28058863 19.446851,6.33132057 19.4854124,6.39789556 C19.8110278,6.97418137 19.9768461,7.69087679 19.9768461,8.56681414 C19.9768461,9.23824368 19.8649625,9.84134739 19.638078,10.3740328 C19.4122611,10.9065473 19.1375912,11.3415267 18.817143,11.6765369 C18.496652,12.0127855 18.1752643,12.3201237 17.8538766,12.5967152 C17.5324462,12.8739047 17.2577763,13.162923 17.0328135,13.4629589 C16.8060571,13.7620125 16.6932767,14.0712297 16.6932767,14.3917206 C16.6932767,14.7126813 16.8397076,15.0387237 17.1317151,15.3673283 C17.4226551,15.6968724 17.7805972,16.0162103 18.2038758,16.3317049 C18.6279231,16.6450216 19.0510736,16.9928002 19.4743522,17.3712399 C19.8985703,17.74921 20.2545052,18.2354765 20.5454452,18.8259399 C20.8385631,19.4186239 20.9848231,20.071093 20.9848231,20.7882582 C20.9848231,21.7343577 20.7435475,22.5887297 20.2626616,23.3491963 C19.7802812,24.1064175 19.1518542,24.7106315 18.3796867,25.1537246 C17.6054695,25.599807 16.777531,25.9355858 15.8945476,26.1623422 C15.0097279,26.3870488 14.1317408,26.5 13.2548639,26.5 C12.7014246,26.5 12.1429889,26.4573818 11.5814359,26.3697112 C11.0178757,26.2822114 10.4532907,26.1276668 9.8847343,25.9098782 C9.31502485,25.6906377 8.81052389,25.4214338 8.37319577,25.0981672 C7.93475736,24.778146 7.58181166,24.3652016 7.3111559,23.862238 C7.04054285,23.3591036 6.90615445,22.7936645 6.90615445,22.1655364 C6.90615445,21.4203577 7.1136515,20.7291563 7.52975591,20.0847582 C7.94586031,19.4451002 8.49707907,18.9108775 9.18272893,18.487727 C10.3789864,17.7434877 12.2558547,17.2836547 14.8102164,17.1098936 C14.2262867,16.3799173 13.9333396,15.6928582 13.9333396,15.0498693 C13.9333396,14.6841125 14.0289531,14.2920503 14.218173,13.8687717 C13.9130554,13.9113899 13.5987992,13.9355175 13.2781801,13.9355175 C11.9059836,13.9355175 10.7472627,13.4894351 9.80628769,12.5911638 C8.86535538,11.6944724 8.39535897,10.5719642 8.39535897,9.21488479 C8.39535897,9.07298054 8.39941581,8.94802965 8.40970739,8.8095417 L2.83316537,8.8095417 L11.2417283,1.5 L25.1668346,1.5 L23.87996,2.50451805 L23.87996,4.27748834 C24.3577712,4.3388962 24.3072955,4.62607827 24.3072955,4.96950105 L24.3072528,4.96950105 Z M15.8172967,17.9839947 C15.6566455,17.9549562 15.430914,17.9386434 15.1399313,17.9386434 C14.512273,17.9386434 13.8947782,17.9941154 13.2885571,18.1048887 C12.6822934,18.2126727 12.084186,18.3947184 11.4941496,18.6515809 C10.9019353,18.906906 10.423996,19.2823566 10.0597766,19.7787865 C9.69355004,20.274576 9.51150436,20.858463 9.51150436,21.5303623 C9.51150436,22.1706609 9.67215551,22.7419077 9.99371402,23.2394479 C10.3142049,23.7338281 10.7373127,24.1208513 11.2632509,24.3980408 C11.7891891,24.6756999 12.3405787,24.8854176 12.916224,25.0231795 C13.4929795,25.1600875 14.0799412,25.2311463 14.6780913,25.2311463 C15.8610679,25.2311463 16.8794647,24.9647609 17.7338794,24.4320756 C18.5863724,23.899561 19.0135798,23.0772595 19.0135798,21.9672634 C19.0135798,21.7338025 18.9810823,21.5040141 18.916856,21.2796064 C18.8498112,21.0527219 18.7835352,20.8585057 18.7185829,20.6972567 C18.6534171,20.5389543 18.529363,20.3483679 18.3462498,20.1286576 C18.1642041,19.9083067 18.0248194,19.7455631 17.9310848,19.6372239 C17.8354714,19.5253403 17.6565431,19.3656714 17.3929761,19.1545445 C17.131587,18.9425209 16.9647012,18.8103104 16.8904822,18.7614574 C16.8172027,18.7100422 16.6270434,18.5711699 16.3207728,18.3458227 C16.0147157,18.1186393 15.8468903,17.9981723 15.8173394,17.983952 L15.8172967,17.9839947 Z"></path>
        </svg>
    </a>
    

    

    
    <a class="social-icon" href="https://orcid.org/0000-0001-6635-5531" target="_blank" rel="noopener" title="Orcid">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M27,14 C27,21.1804687 21.1804687,27 14,27 C6.81953125,27 1,21.1804687 1,14 C1,6.81953125 6.81953125,1 14,1 C21.1804687,1 27,6.81953125 27,14 Z M9.76484375,19.9109375 L9.76484375,13.9492188 L9.76484375,9.03359375 L8.20078125,9.03359375 L8.20078125,19.9109375 L9.76484375,19.9109375 Z M12.0601563,9.03359375 L12.0601563,19.9210938 L16.3054688,19.9210938 C19.890625,19.9210938 22.0742188,17.2703125 22.0742188,14.4773437 C22.0742188,11.9078125 20.3070313,9.03359375 16.2851563,9.03359375 L12.0601563,9.03359375 Z M13.6242188,18.509375 L13.6242188,10.4453125 L16.03125,10.4453125 C19.078125,10.4453125 20.4695313,12.29375 20.4695313,14.4773437 C20.4695313,15.8179688 19.6570313,18.509375 16.1125,18.509375 L13.6242188,18.509375 Z M10.0085938,6.76875 C10.0085938,6.21015625 9.5515625,5.74296875 8.9828125,5.74296875 C8.4140625,5.74296875 7.95703125,6.2 7.95703125,6.76875 C7.95703125,7.32734375 8.4140625,7.79453125 8.9828125,7.79453125 C9.5515625,7.79453125 10.0085938,7.32734375 10.0085938,6.76875 Z"></path>
        </svg>
    </a>
    

    

    

    

</div>




    <p><a href="https://github.com/kimcc/hugo-theme-noteworthy" target="_blank" rel="noopener">Noteworthy theme</a></p>
    <p><a href="https://gohugo.io" target="_blank" rel="noopener">Built with Hugo</a></p>

	<script src="https://lexinwei.github.io/js/main.min.fa5c2b23e07b5d9bfad267a52b4b24fdb053e6fb7524993383594926a3ac270c.js" integrity="sha256-+lwrI+B7XZv60melK0sk/bBT5vt1JJkzg1lJJqOsJww=" crossorigin="anonymous"></script>
</footer>

</nav>

        <div id="content" class="content-container">
        

<h1 class="post-title"></h1>
    
    <div>
        <p>
        <h1 id="k-mer-analysis"><em>K</em>-mer Analysis</h1>
<p>Wei Xin in JiangLab, Oct, 2020</p>
<h2 id="k-mer"><em>k</em>-mer</h2>
<blockquote>
<p>In bioinformatics, <em>k</em>-mers are subsequences of length <em><strong>k</strong></em> contained within a biological sequence. <em>K</em>-mers are composed of nucleotides (i.e. A, T, G, and C). [<a href="https://en.wikipedia.org/wiki/K-mer">k-mer - Wikipedia</a>]</p>
</blockquote>
<h2 id="application">Application</h2>
<blockquote>
<ul>
<li>Sequence assembly</li>
<li>The detection of genetic islands associated with pathogenicity (<strong>dinucleotide</strong> bias)</li>
<li>Genomics-based taxonomy (e.g. <strong>GC-content</strong>)</li>
<li>Identify species in metagenomic samples</li>
<li>Improve heterologous gene expression (<strong>codon</strong> bias)</li>
<li>Create attenuated vaccines (<strong>codon-pair</strong> manipulation)</li>
</ul>
</blockquote>
<p>The frequency of k-mers can be used as a &ldquo;signature&rdquo; of sequence.</p>
<h2 id="some-homemade-tools">Some homemade tools</h2>
<p>Recently, I wrote several custom functions that can be used for basic sequence processing and <em>k</em>-mers analysis for based on <code>python3</code>.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>base<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 ~<span style="color:#f92672">]</span>$ cd kmer/script/mytools/
<span style="color:#f92672">(</span>base<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ls
ATGCcontent  DINULC       KmerSpectra  TRANSLATE
COMPLEMENT   KmerCounter  LENGTH       TRINULC
<span style="color:#f92672">(</span>base<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ 
</code></pre></div><p>The name of each tool corresponds to a Python script and can be invoked independently by <code>shell</code> commands.</p>
<h3 id="get_fasta">get_fasta()</h3>
<p>Load FASTA data. The following tools (functions) all require <code>get_fasta()</code> to read the <code>.fasta</code> file first. This function can read the <code>.fasta</code> file line by line and save the head and sequence as <code>key</code> and <code>value</code> in a dictionary.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#66d9ef">def</span> <span style="color:#a6e22e">get_fasta</span>(fasta_path):
    fasta <span style="color:#f92672">=</span> {}
    <span style="color:#66d9ef">print</span>(<span style="color:#e6db74">&#34;</span><span style="color:#ae81ff">\n</span><span style="color:#e6db74">Getting DNA sequence...&#34;</span>)
    <span style="color:#66d9ef">with</span> open(fasta_path, <span style="color:#e6db74">&#39;r&#39;</span>) <span style="color:#66d9ef">as</span> file:
	    <span style="color:#66d9ef">for</span> line <span style="color:#f92672">in</span> file:
	        <span style="color:#66d9ef">if</span> line<span style="color:#f92672">.</span>startswith(<span style="color:#e6db74">&#34;&gt;&#34;</span>):
	            name <span style="color:#f92672">=</span> line<span style="color:#f92672">.</span>rstrip()
	            fasta[name] <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;&#39;</span>
	            <span style="color:#66d9ef">continue</span>
	        fasta[name] <span style="color:#f92672">+=</span> line<span style="color:#f92672">.</span>rstrip()
    <span style="color:#66d9ef">print</span>(<span style="color:#e6db74">&#34;There are </span><span style="color:#e6db74">%d</span><span style="color:#e6db74"> sequences.</span><span style="color:#ae81ff">\n</span><span style="color:#e6db74">&#34;</span> <span style="color:#f92672">%</span> len(fasta))
    <span style="color:#66d9ef">return</span>(fasta)
</code></pre></div><h3 id="length">LENGTH</h3>
<p>Count DNA bases, i.e. get sequence length.</p>
<h4 id="parameters">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
</tbody>
</table>
<h4 id="examples">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./LENGTH -h
usage: LENGTH <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH

Count DNA bases, i.e. get sequence length.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
  
<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./LENGTH -p /public/home/weixin/kmer/fasta/EcoliK12.fasta 

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Counting DNA bases...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
length: <span style="color:#ae81ff">4641652</span> bases
</code></pre></div><p>The FASTA file contains one sequence of length 4,641,652.</p>
<h3 id="complement">COMPLEMENT</h3>
<p>Get DNA complementary strand.</p>
<h4 id="parameters-1">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
<tr>
<td>-r, &ndash;reverse</td>
<td>whether to reverse complement (default False)</td>
</tr>
<tr>
<td>-w, &ndash;width</td>
<td>the number of bases per line output to FASTA (default 60)</td>
</tr>
<tr>
<td>-o, &ndash;outpath</td>
<td>specify the output file (default is to FASTA directory)</td>
</tr>
</tbody>
</table>
<h4 id="examples-1">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./COMPLEMENT -h
usage: COMPLEMENT <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH <span style="color:#f92672">[</span>--reverse REVERSE<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--width WIDTH<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--outpath OUTPATH<span style="color:#f92672">]</span>

Get DNA complementary strand.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
  --reverse REVERSE, -r REVERSE
                        whether to reverse complement <span style="color:#f92672">(</span>default False<span style="color:#f92672">)</span>
  --width WIDTH, -w WIDTH
                        the number of bases per line output to FASTA <span style="color:#f92672">(</span>default 60<span style="color:#f92672">)</span>
  --outpath OUTPATH, -o OUTPATH
                        specify the output file <span style="color:#f92672">(</span>default is to FASTA directory<span style="color:#f92672">)</span>

<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./COMPLEMENT -p /public/home/weixin/kmer/fasta/EcoliK12.fasta -r Ture -w <span style="color:#ae81ff">60</span> -o /public/home/weixin/kmer/fasta/EcoliK12_reverse_complement.fasta

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Getting complementary strand...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
OK :<span style="color:#f92672">)</span>
The FASTA file of the complementary sequence has been saved in <span style="color:#e6db74">&#39;/public/home/weixin/kmer/fasta/EcoliK12_reverse_complement.fasta&#39;</span>.

<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ head -5 /public/home/weixin/kmer/fasta/EcoliK12_reverse_complement.fasta
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
GAAAAATACTTACTAAGGCGTTTTTTATTTGGTGATATTTTTTTCAATATCATGCAGCAA
ACGGTGCAACATTGCCGTGTCTCGTTGCTCTAAAAGCCCCAGGCGTTGTTGTAACCAGTC
GACCAGTTTTATGTCATCTGCCACTGCCAGAGTCGTCAGCAATGTCATGGCTCGTTCGCG
TAAAGCTTGCAGTTGATGTTGGTCTGCCGTTGCATCACTTTTCGCCGGTTGTTGTATTAA
</code></pre></div><h3 id="translate">TRANSLATE</h3>
<h4 id="parameters-2">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
<tr>
<td>-w, &ndash;width</td>
<td>set the number of AA per line output to FASTA (default 60)</td>
</tr>
<tr>
<td>-b, &ndash;begin</td>
<td>set translation start location (generally 1/2/3, default 1)</td>
</tr>
<tr>
<td>-o, &ndash;outpath</td>
<td>specify the output file (default is to FASTA directory)</td>
</tr>
</tbody>
</table>
<h4 id="examples-2">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./TRANSLATE -h
usage: TRANSLATE <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH <span style="color:#f92672">[</span>--width WIDTH<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--begin BEGIN<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--outpath OUTPATH<span style="color:#f92672">]</span>

Translate DNA into AA sequence.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
  --width WIDTH, -w WIDTH
                        set the number of AA per line output to FASTA <span style="color:#f92672">(</span>default 60<span style="color:#f92672">)</span>
  --begin BEGIN, -b BEGIN
                        set translation start location <span style="color:#f92672">(</span>generally 1/2/3, default 1<span style="color:#f92672">)</span>
  --outpath OUTPATH, -o OUTPATH
                        specify the output file <span style="color:#f92672">(</span>default is to FASTA directory<span style="color:#f92672">)</span>

<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./TRANSLATE -p /public/home/weixin/kmer/fasta/EcoliK12.fasta -w <span style="color:#ae81ff">60</span> -b <span style="color:#ae81ff">1</span> -o /public/home/weixin/kmer/fasta/EcoliK12_translated.fasta

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Translating DNA into AA sequence...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
OK :<span style="color:#f92672">)</span>
The FASTA files of the AA sequence has been saved in <span style="color:#e6db74">&#39;/public/home/weixin/kmer/fasta/EcoliK12_translated.fasta&#39;</span>.

<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ head -5 /public/home/weixin/kmer/fasta/EcoliK12_translated.fasta
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
SFSF_LQRAICLCVD_KKSV__QLLNWLPAVSKLKFY_LRSLNTLTNIGIAHRQIKITEY
TTSMKRISTTITTTITITTGNGAG_RVQETQKKART_QCGLFFSTKGNEVTTMRVLKFGG
TSVANAERFLRVADILESNARQGQVATVLSAPAKITNHLVAMIEKTISGQDALPNISDAE
RIFAELLTGLAAAQPGFPLAQLKTFVDQEFAQIKHVLHGISLLGQCPDSINAALICRGEK
</code></pre></div><h3 id="atgccontent">ATGCcontent</h3>
<h4 id="description">Description</h4>
<p>Count A|T|G|C (1-mers) of DNA sequence (N - unknown base).</p>
<h4 id="parameter">Parameter</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
</tbody>
</table>
<h4 id="examples-3">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./ATGCcontent -h
usage: ATGCcontent <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH

Count A|T|G|C of DNA sequence <span style="color:#f92672">(</span>N - unknown base<span style="color:#f92672">)</span>.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file

<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./ATGCcontent -p /public/home/weixin/kmer/fasta/EcoliK12.fasta 

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Counting A|T|G|C of DNA sequence...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
length: <span style="color:#ae81ff">4641652</span> bases
A: <span style="color:#ae81ff">1142742</span> <span style="color:#f92672">(</span>24.6%<span style="color:#f92672">)</span>
T: <span style="color:#ae81ff">1141382</span> <span style="color:#f92672">(</span>24.6%<span style="color:#f92672">)</span>
G: <span style="color:#ae81ff">1177437</span> <span style="color:#f92672">(</span>25.4%<span style="color:#f92672">)</span>
C: <span style="color:#ae81ff">1180091</span> <span style="color:#f92672">(</span>25.4%<span style="color:#f92672">)</span>
ATGC: <span style="color:#ae81ff">4641652</span> <span style="color:#f92672">(</span>100.0%<span style="color:#f92672">)</span>
N: <span style="color:#ae81ff">0</span> <span style="color:#f92672">(</span>0.0%<span style="color:#f92672">)</span>
</code></pre></div><p>In this sequence, A/T/G/C accounts for about a quarter, with slightly higher GC content and no unknown nucleotides.</p>
<h3 id="dinulc">DINULC</h3>
<h4 id="description-1">Description</h4>
<p>Count 2-mers (dinucleotide).</p>
<h4 id="parameters-3">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
</tbody>
</table>
<h4 id="examples-4">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./DINULC -h
usage: DINULC <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH

Count 2-mers.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./DINULC -p /public/home/weixin/kmer/fasta/EcoliK12.fasta 

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Counting 2mers...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
   kmer   count      freq
<span style="color:#ae81ff">1</span>    GC  <span style="color:#ae81ff">384102</span>  8.275116
<span style="color:#ae81ff">11</span>   CG  <span style="color:#ae81ff">346793</span>  7.471329
<span style="color:#ae81ff">3</span>    TT  <span style="color:#ae81ff">339584</span>  7.316018
<span style="color:#ae81ff">10</span>   AA  <span style="color:#ae81ff">338006</span>  7.282021
<span style="color:#ae81ff">5</span>    CA  <span style="color:#ae81ff">325327</span>  7.008864
<span style="color:#ae81ff">7</span>    TG  <span style="color:#ae81ff">322379</span>  6.945352
<span style="color:#ae81ff">6</span>    AT  <span style="color:#ae81ff">309950</span>  6.677581
<span style="color:#ae81ff">15</span>   CC  <span style="color:#ae81ff">271821</span>  5.856127
<span style="color:#ae81ff">12</span>   GG  <span style="color:#ae81ff">270252</span>  5.822325
<span style="color:#ae81ff">4</span>    TC  <span style="color:#ae81ff">267395</span>  5.760773
<span style="color:#ae81ff">8</span>    GA  <span style="color:#ae81ff">267384</span>  5.760536
<span style="color:#ae81ff">9</span>    AC  <span style="color:#ae81ff">256773</span>  5.531932
<span style="color:#ae81ff">14</span>   GT  <span style="color:#ae81ff">255699</span>  5.508794
<span style="color:#ae81ff">0</span>    AG  <span style="color:#ae81ff">238013</span>  5.127766
<span style="color:#ae81ff">2</span>    CT  <span style="color:#ae81ff">236149</span>  5.087608
<span style="color:#ae81ff">13</span>   TA  <span style="color:#ae81ff">212024</span>  4.567857

OK :<span style="color:#f92672">)</span>
</code></pre></div><h3 id="trinulc">TRINULC</h3>
<h4 id="description-2">Description</h4>
<p>Count 3-mers (trinucleotide).</p>
<h4 id="parameters-4">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
</tbody>
</table>
<h4 id="examples-5">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./TRINULC -h
usage: TRINULC <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH

Count 3-mers.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./TRINULC -p /public/home/weixin/kmer/fasta/EcoliK12.fasta 

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Counting 3mers...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
   kmer   count      freq
<span style="color:#ae81ff">57</span>  CGC  <span style="color:#ae81ff">115734</span>  2.493381
<span style="color:#ae81ff">56</span>  GCG  <span style="color:#ae81ff">114670</span>  2.470458
<span style="color:#ae81ff">3</span>   TTT  <span style="color:#ae81ff">109862</span>  2.366874
<span style="color:#ae81ff">34</span>  AAA  <span style="color:#ae81ff">108964</span>  2.347527
<span style="color:#ae81ff">40</span>  CAG  <span style="color:#ae81ff">104850</span>  2.258895
..  ...     ...       ...
<span style="color:#ae81ff">19</span>  GGG   <span style="color:#ae81ff">47515</span>  1.023666
<span style="color:#ae81ff">27</span>  CTC   <span style="color:#ae81ff">42746</span>  0.920923
<span style="color:#ae81ff">37</span>  GAG   <span style="color:#ae81ff">42503</span>  0.915687
<span style="color:#ae81ff">39</span>  TAG   <span style="color:#ae81ff">27254</span>  0.587162
<span style="color:#ae81ff">54</span>  CTA   <span style="color:#ae81ff">26770</span>  0.576735

<span style="color:#f92672">[</span><span style="color:#ae81ff">64</span> rows x <span style="color:#ae81ff">3</span> columns<span style="color:#f92672">]</span>

OK :<span style="color:#f92672">)</span>
</code></pre></div><h3 id="kmercounter">KmerCounter</h3>
<h4 id="description-3">Description</h4>
<p>Count <em>k</em>-mers.</p>
<h4 id="parameters-5">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
<tr>
<td>-k, &ndash;k</td>
<td>k-mer&rsquo;s k (default 8)</td>
</tr>
<tr>
<td>-l, &ndash;label</td>
<td>customize a label used to distinguish the results (the label will become legend in the k-mer spectrum</td>
</tr>
<tr>
<td>-s, &ndash;savetofile</td>
<td>whether to save the count for each k-mer to a text file (default True)</td>
</tr>
<tr>
<td>-o, &ndash;outdir</td>
<td>specify the output directory (default is to FASTA directory)</td>
</tr>
</tbody>
</table>
<h4 id="examples-6">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./KmerCounter -h
usage: KmerCounter <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH <span style="color:#f92672">[</span>--k K<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--label LABEL<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--savetofile SAVETOFILE<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--outdir OUTDIR<span style="color:#f92672">]</span>

Count k-mers.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
  --k K, -k K           k-mer<span style="color:#e6db74">&#39;s k (default 8)
</span><span style="color:#e6db74">  --label LABEL, -l LABEL
</span><span style="color:#e6db74">                        customize a label used to distinguish the results (the label will become legend in the k-mer spectrum）
</span><span style="color:#e6db74">  --savetofile SAVETOFILE, -s SAVETOFILE
</span><span style="color:#e6db74">                        whether to save the count for each k-mer to a text file (default True)
</span><span style="color:#e6db74">  --outdir OUTDIR, -o OUTDIR
</span><span style="color:#e6db74">                        specify the output directory (default is to FASTA directory)
</span><span style="color:#e6db74">                        
</span><span style="color:#e6db74">(wx) [weixin@tc6000 mytools]$ ./KmerCounter -p /public/home/weixin/kmer/fasta/EcoliK12.fasta -k 8 -l EcoliK12 -s True -o /public/home/weixin/kmer
</span><span style="color:#e6db74">
</span><span style="color:#e6db74">Getting DNA sequence...
</span><span style="color:#e6db74">There are 1 sequences.
</span><span style="color:#e6db74">
</span><span style="color:#e6db74">Counting 8mers...
</span><span style="color:#e6db74">&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
</span><span style="color:#e6db74">           kmer  count      freq     label
</span><span style="color:#e6db74">580    CGCTGGCG    778  0.016761  EcoliK12
</span><span style="color:#e6db74">23909  CGCCAGCG    734  0.015813  EcoliK12
</span><span style="color:#e6db74">13626  CCAGCGCC    726  0.015641  EcoliK12
</span><span style="color:#e6db74">8748   CGCCAGCA    688  0.014822  EcoliK12
</span><span style="color:#e6db74">24071  CCGCCAGC    663  0.014284  EcoliK12
</span><span style="color:#e6db74">...         ...    ...       ...       ...
</span><span style="color:#e6db74">63192  TTCCAAGG      1  0.000022  EcoliK12
</span><span style="color:#e6db74">65279  TTCCTAGA      1  0.000022  EcoliK12
</span><span style="color:#e6db74">65163  TTCCTAGT      1  0.000022  EcoliK12
</span><span style="color:#e6db74">65338  TTCTAGAT      1  0.000022  EcoliK12
</span><span style="color:#e6db74">64734  TTGACTAG      1  0.000022  EcoliK12
</span><span style="color:#e6db74">
</span><span style="color:#e6db74">[65360 rows x 4 columns]
</span><span style="color:#e6db74">
</span><span style="color:#e6db74">The results of 8-mer counts has been saved in &#39;</span>/public/home/weixin/kmer/EcoliK12_8mer_seq1.txt<span style="color:#960050;background-color:#1e0010">&#39;</span>
OK :<span style="color:#f92672">)</span>

<span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ head -5 /public/home/weixin/kmer/EcoliK12_8mer_seq1.txt
kmer	count	freq	label
CGCTGGCG	778	0.016761	EcoliK12
CGCCAGCG	734	0.015813	EcoliK12
CCAGCGCC	726	0.015641	EcoliK12
CGCCAGCA	688	0.014822	EcoliK12
</code></pre></div><h3 id="kmerspectra">KmerSpectra</h3>
<h4 id="description-4">Description</h4>
<p>Count <em>k</em>-mers.</p>
<h4 id="parameters-6">Parameters</h4>
<table>
<thead>
<tr>
<th>Arguments</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, &ndash;help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>-p, -path</td>
<td>path of your fasta file</td>
</tr>
<tr>
<td>-k, &ndash;k</td>
<td>k-mer&rsquo;s k (default 8)</td>
</tr>
<tr>
<td>-l, &ndash;label</td>
<td>customize a label used to distinguish the results (the label will become legend in the k-mer spectrum</td>
</tr>
<tr>
<td>-s, &ndash;savetofile</td>
<td>whether to save the count for each k-mer to a text file (default True)</td>
</tr>
<tr>
<td>-o, &ndash;outdir</td>
<td>specify the output directory (default is to FASTA directory)</td>
</tr>
<tr>
<td>-pdf, &ndash;savetopdf</td>
<td>whether to save the spectrum to pdf (default True)</td>
</tr>
<tr>
<td>-pp, &ndash;pdfpath</td>
<td>specify the output directory of k-mer spectrum (default is to current working directory)</td>
</tr>
<tr>
<td>-t, &ndash;title</td>
<td>specify the title of k-mer spectrum</td>
</tr>
<tr>
<td>-c, &ndash;col</td>
<td>specify the colors of points</td>
</tr>
</tbody>
</table>
<h4 id="examples-7">Examples</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ ./KmerSpectra  -h
usage: KmerSpectra <span style="color:#f92672">[</span>-h<span style="color:#f92672">]</span> --path PATH <span style="color:#f92672">[</span>--k K<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--label LABEL<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--savetofile SAVETOFILE<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--outdir OUTDIR<span style="color:#f92672">]</span>
                   <span style="color:#f92672">[</span>--savetopdf SAVETOPDF<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--pdfpath PDFPATH<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--title TITLE<span style="color:#f92672">]</span> <span style="color:#f92672">[</span>--col COL<span style="color:#f92672">]</span>

Count k-mers.

optional arguments:
  -h, --help            show this help message and exit
  --path PATH, -p PATH  path of your fasta file
  --k K, -k K           k-mer<span style="color:#e6db74">&#39;s k (default 8)
</span><span style="color:#e6db74">  --label LABEL, -l LABEL
</span><span style="color:#e6db74">                        customize a label used to distinguish the results (the label will become legend in the k-mer spectrum）
</span><span style="color:#e6db74">  --savetofile SAVETOFILE, -s SAVETOFILE
</span><span style="color:#e6db74">                        whether to save the count for each k-mer to a text file (default True)
</span><span style="color:#e6db74">  --outdir OUTDIR, -o OUTDIR
</span><span style="color:#e6db74">                        specify the output directory of k-mer counts result (default is to FASTA directory)
</span><span style="color:#e6db74">  --savetopdf SAVETOPDF, -pdf SAVETOPDF
</span><span style="color:#e6db74">                        whether to save the spectrum to pdf (default True)
</span><span style="color:#e6db74">  --pdfpath PDFPATH, -pp PDFPATH
</span><span style="color:#e6db74">                        specify the output directory of k-mer spectrum (default is to current working directory)
</span><span style="color:#e6db74">  --title TITLE, -t TITLE
</span><span style="color:#e6db74">                        specify the title of k-mer spectrum
</span><span style="color:#e6db74">  --col COL, -c COL     specify the colors of points
</span><span style="color:#e6db74"> 
</span><span style="color:#e6db74">(wx) [weixin@tc6000 mytools]$ ./KmerSpectra -p /public/home/weixin/kmer/fasta/EcoliK12.fasta -k 8 -l EcoliK12 -s True -o /public/home/weixin/kmer -pdf True -pp /public/home/weixin/kmer/pdf/EcoliK12_8mers_spectra.pdf -t &#39;</span>EcoliK12 8mers spectra<span style="color:#e6db74">&#39; -c &#39;</span><span style="color:#75715e">#1f77b4&#39;</span>

Getting DNA sequence...
There are <span style="color:#ae81ff">1</span> sequences.

Counting 8mers...
&gt;gi|556503834|ref|NC_000913.3| Escherichia coli str. K-12 substr. MG1655, complete genome
           kmer  count      freq     label
<span style="color:#ae81ff">580</span>    CGCTGGCG    <span style="color:#ae81ff">778</span>  0.016761  EcoliK12
<span style="color:#ae81ff">23909</span>  CGCCAGCG    <span style="color:#ae81ff">734</span>  0.015813  EcoliK12
<span style="color:#ae81ff">13626</span>  CCAGCGCC    <span style="color:#ae81ff">726</span>  0.015641  EcoliK12
<span style="color:#ae81ff">8748</span>   CGCCAGCA    <span style="color:#ae81ff">688</span>  0.014822  EcoliK12
<span style="color:#ae81ff">24071</span>  CCGCCAGC    <span style="color:#ae81ff">663</span>  0.014284  EcoliK12
...         ...    ...       ...       ...
<span style="color:#ae81ff">63192</span>  TTCCAAGG      <span style="color:#ae81ff">1</span>  0.000022  EcoliK12
<span style="color:#ae81ff">65279</span>  TTCCTAGA      <span style="color:#ae81ff">1</span>  0.000022  EcoliK12
<span style="color:#ae81ff">65163</span>  TTCCTAGT      <span style="color:#ae81ff">1</span>  0.000022  EcoliK12
<span style="color:#ae81ff">65338</span>  TTCTAGAT      <span style="color:#ae81ff">1</span>  0.000022  EcoliK12
<span style="color:#ae81ff">64734</span>  TTGACTAG      <span style="color:#ae81ff">1</span>  0.000022  EcoliK12

<span style="color:#f92672">[</span><span style="color:#ae81ff">65360</span> rows x <span style="color:#ae81ff">4</span> columns<span style="color:#f92672">]</span>

The results of 8-mer counts has been saved in <span style="color:#e6db74">&#39;/public/home/weixin/kmer/EcoliK12_8mer_seq1.txt&#39;</span>
OK :<span style="color:#f92672">)</span>

Ploting Kmer spectrum...
OK :<span style="color:#f92672">)</span>
The Kmer spectra has been saved to <span style="color:#e6db74">&#39;/public/home/weixin/kmer/pdf/EcoliK12_8mers_spectra.pdf&#39;</span>
</code></pre></div><p>An example 8-mer spectrum for E. coli comparing 8-mers' frequency (i.e. multiplicities) with their number of occurrences.</p>
<p><img src="https://tva1.sinaimg.cn/large/0081Kckwly1gk3y86tub3j30yu0ra77r.jpg" alt="image-20201026223929415" style="zoom: 33%;" /></p>
<p>In addition, the <em>k</em>-mer spectrum of two sequences can be drawn in one diagram.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#75715e"># EcoliK12</span>
fasta_path <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;/public/home/weixin/kmer/fasta/EcoliK12.fasta&#39;</span>
fasta <span style="color:#f92672">=</span> get_fasta(fasta_path)
kmer_ecolik12_df <span style="color:#f92672">=</span> KmerCounter(fasta <span style="color:#f92672">=</span> fasta, fasta_path <span style="color:#f92672">=</span> fasta_path, 
                               k<span style="color:#f92672">=</span><span style="color:#ae81ff">8</span>, savetofile<span style="color:#f92672">=</span>True, label<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;EcoliK12&#34;</span>)

<span style="color:#75715e"># EcoliO157</span>
fasta_path <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;/public/home/weixin/kmer/fasta/EcoliO157.fasta&#39;</span>
fasta <span style="color:#f92672">=</span> get_fasta(fasta_path)
kmer_ecolio157_df <span style="color:#f92672">=</span> KmerCounter(fasta <span style="color:#f92672">=</span> fasta, fasta_path <span style="color:#f92672">=</span> fasta_path, 
                                k<span style="color:#f92672">=</span><span style="color:#ae81ff">8</span>, savetofile<span style="color:#f92672">=</span>True, label<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;EcoliO157&#34;</span>)

<span style="color:#75715e"># two Ecoli</span>
<span style="color:#f92672">import</span> pandas <span style="color:#f92672">as</span> pd
kmer_ecoli_df  <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>concat([kmer_ecolik12_df, kmer_ecolio157_df], axis<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>)
KmerSpectra(kmer_ecoli_df, savetopdf<span style="color:#f92672">=</span>True, 
            pdfpath<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;/public/home/weixin/kmer/pdf/Ecoli_8mer_plot.pdf&#39;</span>, 
            title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;8-mer spectrum of ${E.coli}$&#39;</span>)
</code></pre></div><p><img src="https://tva1.sinaimg.cn/large/0081Kckwly1gk3y885tblj30yw0r60w3.jpg" alt="image-20201026224041208" style="zoom:33%;" /></p>
<h3 id="dependence">Dependence</h3>
<p>My conda environment:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-shell" data-lang="shell"><span style="color:#f92672">(</span>wx<span style="color:#f92672">)</span> <span style="color:#f92672">[</span>weixin@tc6000 mytools<span style="color:#f92672">]</span>$ conda list
<span style="color:#75715e"># packages in environment at /public/home/weixin/miniconda2/envs/wx:</span>
<span style="color:#75715e">#</span>
<span style="color:#75715e"># Name                    Version                   Build  Channel</span>
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
ca-certificates           2020.6.20            hecda079_0    conda-forge
certifi                   2020.6.20        py38h924ce5b_2    conda-forge
cycler                    0.10.0                     py_2    conda-forge
dbus                      1.13.16              hb2f20db_0  
expat                     2.2.9                he1b5a44_2    conda-forge
fontconfig                2.13.1            he4413a7_1000    conda-forge
freetype                  2.10.4               he06d7ca_0    conda-forge
glib                      2.66.1               h92f7085_0  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb31296c_0  
icu                       58.2              hf484d3e_1000    conda-forge
jpeg                      9d                   h516909a_0    conda-forge
kiwisolver                1.2.0            py38hbf85e49_1    conda-forge
lcms2                     2.11                 hbd6801e_0    conda-forge
ld_impl_linux-64          2.35                 h769bd43_9    conda-forge
libblas                   3.8.0               17_openblas    conda-forge
libcblas                  3.8.0               17_openblas    conda-forge
libffi                    3.3                  he6710b0_2  
libgcc-ng                 9.3.0               h5dbcf3e_17    conda-forge
libgfortran-ng            9.3.0               he4bcb1c_17    conda-forge
libgfortran5              9.3.0               he4bcb1c_17    conda-forge
libgomp                   9.3.0               h5dbcf3e_17    conda-forge
liblapack                 3.8.0               17_openblas    conda-forge
libopenblas               0.3.10          pthreads_h4812303_5    conda-forge
libpng                    1.6.37               hed695b0_2    conda-forge
libstdcxx-ng              9.3.0               h2ae2ef3_17    conda-forge
libtiff                   4.1.0                hc7e4089_6    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libwebp-base              1.1.0                h516909a_3    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxml2                   2.9.10               hb55368b_3  
lz4-c                     1.9.2                he1b5a44_3    conda-forge
matplotlib                3.3.2            py38h32f6830_1    conda-forge
matplotlib-base           3.3.2            py38h4d1ce4f_1    conda-forge
ncurses                   6.2                  he1b5a44_2    conda-forge
numpy                     1.19.2           py38hf89b668_1    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
openssl                   1.1.1h               h516909a_0    conda-forge
pandas                    1.1.3                    pypi_0    pypi
pcre                      8.44                 he1b5a44_0    conda-forge
pillow                    8.0.0            py38h9776b28_0    conda-forge
pip                       20.2.4                     py_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyqt                      5.9.2            py38h05f1152_4  
python                    3.8.3                hcff3b4d_2  
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.8                      1_cp38    conda-forge
pytz                      2020.1                   pypi_0    pypi
qt                        5.9.7                h5867ecd_1  
readline                  8.0                  he28a2e2_2    conda-forge
scipy                     1.5.3                    pypi_0    pypi
seaborn                   0.11.0                   pypi_0    pypi
setuptools                49.6.0           py38h924ce5b_2    conda-forge
sip                       4.19.13          py38he6710b0_0  
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sqlite                    3.33.0               h4cf870e_1    conda-forge
tk                        8.6.10               hed695b0_1    conda-forge
tornado                   6.0.4            py38h1e0a361_2    conda-forge
wheel                     0.35.1             pyh9f0ad1d_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.4.5                h6597ccf_2    conda-forge
</code></pre></div><h2 id="downstream-analysis-based-on-k-mer-frequency">Downstream analysis based on <em>k</em>-mer frequency</h2>
<p>Using the above <em>k</em>-mer related tools, I can process <em>E. coli</em> K12, <em>E. coli</em> O157, <em>C. crescentus</em> NA1000 genomes and <em>H. sapiens</em> chr21 sequences respectively. Based on the <em>k</em>-mer counts list output from <code>.KmerCounter</code>, we can analyze the differences and connections between sequences. I used <code>R</code> to do the analysis:</p>
<ul>
<li>Load packages</li>
</ul>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#a6e22e">library</span>(magrittr)
<span style="color:#a6e22e">library</span>(ggplot2)
<span style="color:#a6e22e">library</span>(ggpubr)
<span style="color:#a6e22e">library</span>(reshape2)
<span style="color:#a6e22e">library</span>(factoextra)
<span style="color:#a6e22e">library</span>(FactoMineR)
<span style="color:#a6e22e">library</span>(ggthemes)
</code></pre></div><h3 id="data-prepare">Data prepare</h3>
<h4 id="read-data">Read data</h4>
<p>Take the 8-MER analysis as an example：</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">#### read data ####</span>
ekdf <span style="color:#f92672">=</span> <span style="color:#a6e22e">read.delim2</span>(file <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;./tables/EcoliK12_8mer_seq1.txt&#39;</span>)
eodf <span style="color:#f92672">=</span> <span style="color:#a6e22e">read.delim2</span>(file <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;./tables/EcoliO157_8mer_seq1.txt&#39;</span>)
cndf <span style="color:#f92672">=</span> <span style="color:#a6e22e">read.delim2</span>(file <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;./tables/caulobacterNA1000_8mer_seq1.txt&#39;</span>)
hsdf <span style="color:#f92672">=</span> <span style="color:#a6e22e">read.delim2</span>(file <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;./tables/HomeSapiens21_8mer_seq1.txt&#39;</span>)
<span style="color:#a6e22e">dim</span>(hsdf)
hsdf <span style="color:#f92672">=</span> hsdf[<span style="color:#f92672">!</span><span style="color:#a6e22e">grepl</span>(<span style="color:#e6db74">&#34;N&#34;</span>, hsdf<span style="color:#f92672">$</span>kmer), ]
<span style="color:#a6e22e">dim</span>(hsdf) <span style="color:#75715e">## [1] 65560     4</span>
hsdf<span style="color:#f92672">$</span>freq <span style="color:#f92672">=</span> hsdf<span style="color:#f92672">$</span>count<span style="color:#f92672">/</span><span style="color:#a6e22e">sum</span>(hsdf<span style="color:#f92672">$</span>count) <span style="color:#f92672">*</span> <span style="color:#ae81ff">100</span>
</code></pre></div><p>Approximately 14.3% of the <em>H.sapiens</em> chr21 sequence is unknown base (N). Therefore, 8-mers containing &ldquo;N&rdquo; was eliminated.</p>
<h4 id="merge-data">Merge data</h4>
<p>Take the <code>kmer</code> and <code>count</code> columns from each of the four tables, and merge them into five columns, the first of which is the k-mers after the union.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">#### merge data ####</span>
selcol <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;kmer&#39;</span>, <span style="color:#e6db74">&#39;count&#39;</span>)
mergedf1 <span style="color:#f92672">=</span> <span style="color:#a6e22e">merge</span>(ekdf[selcol], eodf[selcol], by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;kmer&#39;</span>, all <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>, suffixes <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;.ek&#39;</span>, <span style="color:#e6db74">&#39;.eo&#39;</span>))
mergedf2 <span style="color:#f92672">=</span> <span style="color:#a6e22e">merge</span>(mergedf1, cndf[selcol], by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;kmer&#39;</span>, all <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)
<span style="color:#a6e22e">colnames</span>(mergedf2)<span style="color:#a6e22e">[ncol</span>(mergedf2)] <span style="color:#f92672">%&lt;&gt;%</span> <span style="color:#a6e22e">paste0</span>(., <span style="color:#e6db74">&#39;.cn&#39;</span>)
mergedf2<span style="color:#a6e22e">[is.na</span>(mergedf2)] <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>
mergedf3 <span style="color:#f92672">=</span> <span style="color:#a6e22e">merge</span>(mergedf2, hsdf[selcol], by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;kmer&#39;</span>, all <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)
<span style="color:#a6e22e">colnames</span>(mergedf3)<span style="color:#a6e22e">[ncol</span>(mergedf3)] <span style="color:#f92672">%&lt;&gt;%</span> <span style="color:#a6e22e">paste0</span>(., <span style="color:#e6db74">&#39;.hs&#39;</span>)
mergedf3<span style="color:#a6e22e">[is.na</span>(mergedf3)] <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>
mergedf4 <span style="color:#f92672">=</span> <span style="color:#a6e22e">data.frame</span>(
  kmer <span style="color:#f92672">=</span> mergedf3<span style="color:#f92672">$</span>kmer,
  <span style="color:#a6e22e">apply</span>(mergedf3[, <span style="color:#ae81ff">-1</span>], <span style="color:#ae81ff">2</span>, <span style="color:#a6e22e">function</span>(x){x<span style="color:#f92672">/</span><span style="color:#a6e22e">sum</span>(x)<span style="color:#f92672">*</span><span style="color:#ae81ff">10</span>^5})
)
<span style="color:#a6e22e">dim</span>(mergedf4) <span style="color:#75715e">## [1] 65560     5</span>
</code></pre></div><p>The unique 8-mer of each sequence is ≤ 4<sup>8</sup>, after taking the union, all possible 65,560 8-mers are covered.</p>
<h3 id="counts-distribution">Counts distribution</h3>
<p>Add up the counts of 8-mer from large to small, and draw the accumulation curves.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e"># set colors for plot</span>
mycol <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;#1f77b4&#39;</span>, <span style="color:#e6db74">&#39;#ff7f0e&#39;</span>, <span style="color:#e6db74">&#39;#2ca02c&#39;</span>, <span style="color:#e6db74">&#39;#d62728&#39;</span>, <span style="color:#e6db74">&#39;#9467bd&#39;</span>,
          <span style="color:#e6db74">&#39;#8c564b&#39;</span>, <span style="color:#e6db74">&#39;#e377c2&#39;</span>, <span style="color:#e6db74">&#39;#7f7f7f&#39;</span>, <span style="color:#e6db74">&#39;#bcbd22&#39;</span>, <span style="color:#e6db74">&#39;#17becf&#39;</span>)
<span style="color:#75715e">#### accumulation curve #####</span>
<span style="color:#a6e22e">getwd</span>()
accdf <span style="color:#f92672">=</span> mergedf3[, <span style="color:#ae81ff">-1</span>]
acc <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>
acc.num <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>
cs_df0 <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>
<span style="color:#a6e22e">for </span>(i in <span style="color:#a6e22e">seq_len</span>(<span style="color:#a6e22e">ncol</span>(accdf))) {
  c <span style="color:#f92672">=</span> accdf[, i] <span style="color:#f92672">%&gt;%</span> <span style="color:#a6e22e">sort</span>(., decreasing <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)
  cs_df0 <span style="color:#f92672">=</span> <span style="color:#a6e22e">cbind</span>(cs_df0, <span style="color:#a6e22e">cumsum</span>(c))
}
<span style="color:#a6e22e">colnames</span>(cs_df0) <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;E.coli K12&#39;</span>, <span style="color:#e6db74">&#39;E.coli O157&#39;</span>, <span style="color:#e6db74">&#39;C.crescentus NA1000&#39;</span>, <span style="color:#e6db74">&#39;H.sapiens chr21&#39;</span>)
cs_df1 <span style="color:#f92672">=</span> <span style="color:#a6e22e">data.frame</span>(
  order <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#a6e22e">nrow</span>(cs_df0),
  cs_df0
)
cs_df2 <span style="color:#f92672">=</span> <span style="color:#a6e22e">melt</span>(cs_df1, id.vars <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;order&#39;</span>), variable.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, value.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;count&#39;</span>)
lp1 <span style="color:#f92672">=</span> <span style="color:#a6e22e">ggplot</span>(cs_df2, <span style="color:#a6e22e">aes</span>(x<span style="color:#f92672">=</span>order, y<span style="color:#f92672">=</span>count, color <span style="color:#f92672">=</span> label)) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">geom_line</span>() <span style="color:#f92672">+</span> 
  <span style="color:#a6e22e">xlab</span>(<span style="color:#e6db74">&#39;8-mers number&#39;</span>) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">ylab</span>(<span style="color:#e6db74">&#39;&#39;</span>) <span style="color:#f92672">+</span> 
  <span style="color:#a6e22e">labs</span>(title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Accumulation curves of 8-mer counts&#39;</span>) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">scale_color_manual</span>(values <span style="color:#f92672">=</span> mycol[1<span style="color:#f92672">:</span><span style="color:#ae81ff">4</span>]) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme_few</span>() <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">0.75</span>, <span style="color:#ae81ff">0.50</span>))
<span style="color:#a6e22e">print</span>(lp1)

lp2 <span style="color:#f92672">=</span> <span style="color:#a6e22e">ggplot</span>(cs_df2[cs_df2<span style="color:#f92672">$</span>label <span style="color:#f92672">!=</span> <span style="color:#e6db74">&#39;H.sapiens.chr21&#39;</span>, ], <span style="color:#a6e22e">aes</span>(x<span style="color:#f92672">=</span>order, y<span style="color:#f92672">=</span>count, color <span style="color:#f92672">=</span> label)) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">geom_line</span>() <span style="color:#f92672">+</span> 
  <span style="color:#a6e22e">xlab</span>(<span style="color:#e6db74">&#39;8-mers number&#39;</span>) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">ylab</span>(<span style="color:#e6db74">&#39;&#39;</span>) <span style="color:#f92672">+</span> 
  <span style="color:#a6e22e">labs</span>(title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Accumulation curves of 8-mer counts&#39;</span>) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">scale_color_manual</span>(values <span style="color:#f92672">=</span> mycol[1<span style="color:#f92672">:</span><span style="color:#ae81ff">4</span>]) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme_few</span>() <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">0.75</span>, <span style="color:#ae81ff">0.25</span>))
<span style="color:#a6e22e">print</span>(lp2)
</code></pre></div><p>The sequence of <em>H. Sapiens</em> chr21 is about 10 times the length of these bacterial genomes.</p>
<p><img src="https://tva1.sinaimg.cn/large/0081Kckwly1gk3y84ohiuj31fi0pegpz.jpg" alt="image-20201027095532665" style="zoom:32%;" /></p>
<h3 id="calculate-frequency">Calculate frequency</h3>
<p>To make the 8-mers of the four sequences comparable, we can divide each count by the total 8-mer of the sequence to get the frequency and then multiply by 10<sup>5</sup>.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">#### boxplot ####</span>
bpdf1 <span style="color:#f92672">=</span> <span style="color:#a6e22e">melt</span>(mergedf3, id.vars <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;kmer&#39;</span>), variable.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, value.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;count&#39;</span>)
bpdf1<span style="color:#f92672">$</span>count <span style="color:#f92672">=</span> <span style="color:#a6e22e">log10</span>(bpdf1<span style="color:#f92672">$</span>count<span style="color:#ae81ff">+1</span>)
<span style="color:#a6e22e">range</span>(bpdf1<span style="color:#f92672">$</span>count)
bp1 <span style="color:#f92672">=</span> <span style="color:#a6e22e">ggboxplot</span>(bpdf1, x <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;count&#34;</span>, 
                color <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, bxp.errorbar <span style="color:#f92672">=</span> T,
                bxp.errorbar.width <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.3</span>,
                size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.3</span>, outlier.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>, outlier.shape <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>,
                palette <span style="color:#f92672">=</span> mycol[1<span style="color:#f92672">:</span><span style="color:#ae81ff">4</span>]) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">scale_x_discrete</span>(labels <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;E.coli K12&#39;</span>, <span style="color:#e6db74">&#39;E.coli O157&#39;</span>, <span style="color:#e6db74">&#39;C.crescentus NA1000&#39;</span>, <span style="color:#e6db74">&#39;H.sapiens chr21&#39;</span>)) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;none&#39;</span>, axis.text.x <span style="color:#f92672">=</span> <span style="color:#a6e22e">element_text</span>(angle <span style="color:#f92672">=</span> <span style="color:#ae81ff">30</span>, hjust <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>)) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">labs</span>(x <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;8-mer counts&#39;</span>,
       title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Distributions of 8-mer counts&#39;</span>)
<span style="color:#a6e22e">print</span>(bp1)

<span style="color:#75715e"># calculate frequency</span>
mergedf4 <span style="color:#f92672">=</span> <span style="color:#a6e22e">data.frame</span>(
  kmer <span style="color:#f92672">=</span> mergedf3<span style="color:#f92672">$</span>kmer,
  <span style="color:#a6e22e">apply</span>(mergedf3[, <span style="color:#ae81ff">-1</span>], <span style="color:#ae81ff">2</span>, <span style="color:#a6e22e">function</span>(x){x<span style="color:#f92672">/</span><span style="color:#a6e22e">sum</span>(x)<span style="color:#f92672">*</span><span style="color:#ae81ff">10</span>^5})
)
<span style="color:#a6e22e">dim</span>(mergedf4)

bpdf2 <span style="color:#f92672">=</span> <span style="color:#a6e22e">melt</span>(mergedf4, id.vars <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;kmer&#39;</span>), variable.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, value.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;count&#39;</span>)
bpdf2<span style="color:#f92672">$</span>count <span style="color:#f92672">=</span> <span style="color:#a6e22e">log10</span>(bpdf2<span style="color:#f92672">$</span>count<span style="color:#ae81ff">+1</span>)
<span style="color:#a6e22e">range</span>(bpdf2<span style="color:#f92672">$</span>count)
bp2 <span style="color:#f92672">=</span> <span style="color:#a6e22e">ggboxplot</span>(bpdf2, x <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;count&#34;</span>, 
                color <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;label&#39;</span>, bxp.errorbar <span style="color:#f92672">=</span> T,
                bxp.errorbar.width <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.3</span>,
                size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.3</span>, outlier.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>, outlier.shape <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>,
                palette <span style="color:#f92672">=</span> mycol[1<span style="color:#f92672">:</span><span style="color:#ae81ff">4</span>]) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">scale_x_discrete</span>(labels <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;E.coli K12&#39;</span>, <span style="color:#e6db74">&#39;E.coli O157&#39;</span>, <span style="color:#e6db74">&#39;C.crescentus NA1000&#39;</span>, <span style="color:#e6db74">&#39;H.sapiens chr21&#39;</span>)) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;none&#39;</span>, axis.text.x <span style="color:#f92672">=</span> <span style="color:#a6e22e">element_text</span>(angle <span style="color:#f92672">=</span> <span style="color:#ae81ff">30</span>, hjust <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>)) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">labs</span>(x <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;log10(8-mer frequency*10e5 + 1)&#39;</span>,
       title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Distributions of 8-mer frequency&#39;</span>)
<span style="color:#a6e22e">print</span>(bp2)
</code></pre></div><p><img src="https://tva1.sinaimg.cn/large/0081Kckwly1gk3y85h5rsj31cy0ri439.jpg" alt="image-20201027124447713" style="zoom:33%;" /></p>
<p>In the counts list, we can find that the two highest outliers of <em>H. Sapiens</em> chr21 are corresponding to subsequences poly-T and poly-A, embodying eukaryotic characteristics.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> hsdf <span style="color:#f92672">=</span> hsdf<span style="color:#a6e22e">[order</span>(hsdf<span style="color:#f92672">$</span>count, decreasing <span style="color:#f92672">=</span> T), ]
<span style="color:#f92672">&gt;</span> hsdf[1<span style="color:#f92672">:</span><span style="color:#ae81ff">2</span>,]
      kmer count      freq         label
<span style="color:#ae81ff">2</span> TTTTTTTT <span style="color:#ae81ff">55795</span> <span style="color:#ae81ff">0.1391803</span> HomeSapiens21
<span style="color:#ae81ff">3</span> AAAAAAAA <span style="color:#ae81ff">53466</span> <span style="color:#ae81ff">0.1333706</span> HomeSapiens21
</code></pre></div><p>In addition, compared with <em>E. coli</em>, there are 92 8-mers with count &gt;1000 in <em>C.</em> crescentus .</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">table</span>(eodf<span style="color:#f92672">$</span>count <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1000</span>)

<span style="color:#66d9ef">FALSE</span> 
<span style="color:#ae81ff">65452</span> 
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">table</span>(ekdf<span style="color:#f92672">$</span>count <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1000</span>)

<span style="color:#66d9ef">FALSE</span> 
<span style="color:#ae81ff">65360</span> 
<span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">table</span>(cndf<span style="color:#f92672">$</span>count <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1000</span>)

<span style="color:#66d9ef">FALSE</span>  <span style="color:#66d9ef">TRUE</span> 
<span style="color:#ae81ff">64331</span>    <span style="color:#ae81ff">92</span> 
</code></pre></div><p>It can be seen that most of these sub-sequences are CG bases.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> cndf <span style="color:#f92672">=</span> cndf<span style="color:#a6e22e">[order</span>(cndf<span style="color:#f92672">$</span>count, decreasing <span style="color:#f92672">=</span> T), ]
<span style="color:#f92672">&gt;</span> cndf[cndf<span style="color:#f92672">$</span>count <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1000</span>, ]
       kmer count                 freq             label
<span style="color:#ae81ff">1</span>  CGGCGGCG  <span style="color:#ae81ff">2825</span>  <span style="color:#ae81ff">0.06987520412216709</span> caulobacterNA1000
<span style="color:#ae81ff">2</span>  CGCCGCCG  <span style="color:#ae81ff">2816</span>  <span style="color:#ae81ff">0.06965259285239735</span> caulobacterNA1000
<span style="color:#ae81ff">3</span>  GGCGGCGG  <span style="color:#ae81ff">2163</span> <span style="color:#ae81ff">0.053500908501326515</span> caulobacterNA1000
        ··· ··· omit <span style="color:#ae81ff">86</span> lines here ··· ···
<span style="color:#ae81ff">90</span> GGCGCGGG  <span style="color:#ae81ff">1009</span> <span style="color:#ae81ff">0.024957196799740382</span> caulobacterNA1000
<span style="color:#ae81ff">91</span> GGCGGCCT  <span style="color:#ae81ff">1008</span> <span style="color:#ae81ff">0.024932462214210416</span> caulobacterNA1000
<span style="color:#ae81ff">92</span> GCCCGCGC  <span style="color:#ae81ff">1007</span> <span style="color:#ae81ff">0.024907727628680446</span> caulobacterNA1000
</code></pre></div><h3 id="sequence-similarity">Sequence similarity</h3>
<h4 id="clustering">Clustering</h4>
<p>Based on the normalized frequency, we conducted hierarchical clustering to investigated the similarity between sequences (species).</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#75715e">#### Clustering ####</span>
cdf <span style="color:#f92672">=</span> <span style="color:#a6e22e">scale</span>(<span style="color:#a6e22e">t</span>(mergedf4[<span style="color:#ae81ff">-1</span>]))
<span style="color:#a6e22e">row.names</span>(cdf) <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;E.coliK12&#39;</span>, <span style="color:#e6db74">&#39;E.coliO157&#39;</span>, <span style="color:#e6db74">&#39;C.crescentus NA1000&#39;</span>, <span style="color:#e6db74">&#39;H.sapiens chr21&#39;</span>)
res <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">hcut</span>(cdf, k <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span>, stand <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)
<span style="color:#a6e22e">fviz_dend</span>(res, k<span style="color:#f92672">=</span><span style="color:#ae81ff">4</span>,
          cex <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.5</span>,
          k_colors <span style="color:#f92672">=</span> mycol[2<span style="color:#f92672">:</span><span style="color:#ae81ff">4</span>] <span style="color:#f92672">%&gt;%</span> <span style="color:#a6e22e">rev</span>())
</code></pre></div><p><img src="https://tva1.sinaimg.cn/large/0081Kckwly1gk3y865udij30oq0o4ab8.jpg" alt="image-20201026225431993" style="zoom:33%;" /></p>
<p>Compared with the <em>H.sapiens</em> chr21, the three bacterial genomes are also grouped together in a cluster.</p>
<h4 id="principal-components-analysis-pca">Principal Components Analysis (PCA)</h4>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R">pcadf <span style="color:#f92672">=</span> <span style="color:#a6e22e">t</span>(mergedf4[<span style="color:#ae81ff">-1</span>])
<span style="color:#a6e22e">row.names</span>(pcadf) <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;E.coli K12&#39;</span>, <span style="color:#e6db74">&#39;E.coli O157&#39;</span>, <span style="color:#e6db74">&#39;C.crescentus NA1000&#39;</span>, <span style="color:#e6db74">&#39;H.sapiens chr21&#39;</span>)
res.pca <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">PCA</span>(pcadf,  graph <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>)
<span style="color:#a6e22e">summary</span>(res.pca)
<span style="color:#a6e22e">fviz_pca_ind</span>(res.pca, col.ind <span style="color:#f92672">=</span> <span style="color:#a6e22e">factor</span>(<span style="color:#a6e22e">row.names</span>(pcadf), levels <span style="color:#f92672">=</span> <span style="color:#a6e22e">row.names</span>(pcadf)),
             repel <span style="color:#f92672">=</span> T, geom <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;point&#34;</span>, <span style="color:#e6db74">&#34;text&#34;</span>),
             palette <span style="color:#f92672">=</span> mycol[1<span style="color:#f92672">:</span><span style="color:#ae81ff">4</span>]) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;none&#34;</span>) <span style="color:#f92672">+</span>
  <span style="color:#a6e22e">scale_shape_manual</span>(values <span style="color:#f92672">=</span> <span style="color:#a6e22e">rep</span>(<span style="color:#ae81ff">19</span>, <span style="color:#ae81ff">4</span>))
</code></pre></div><p><img src="https://tva1.sinaimg.cn/large/0081Kckwly1gk3y87qlwdj30r20r0gnz.jpg" alt="image-20201026225351661" style="zoom:33%;" /></p>
<p>As we can see from the figures above, the two strains of <em>E. coli</em> are the most similar.</p>
<h3 id="dependence-1">Dependence</h3>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#f92672">&gt;</span> <span style="color:#a6e22e">sessionInfo</span>()
R version <span style="color:#ae81ff">4.0.2</span> (<span style="color:#ae81ff">2020-06-22</span>)
Platform<span style="color:#f92672">:</span> x86_64<span style="color:#f92672">-</span>apple<span style="color:#f92672">-</span><span style="color:#a6e22e">darwin17.0 </span>(<span style="color:#ae81ff">64</span><span style="color:#f92672">-</span>bit)
Running under<span style="color:#f92672">:</span> macOS Catalina <span style="color:#ae81ff">10.15.7</span>

Matrix products<span style="color:#f92672">:</span> default
BLAS<span style="color:#f92672">:</span>   <span style="color:#f92672">/</span>System<span style="color:#f92672">/</span>Library<span style="color:#f92672">/</span>Frameworks<span style="color:#f92672">/</span>Accelerate.framework<span style="color:#f92672">/</span>Versions<span style="color:#f92672">/</span>A<span style="color:#f92672">/</span>Frameworks<span style="color:#f92672">/</span>vecLib.framework<span style="color:#f92672">/</span>Versions<span style="color:#f92672">/</span>A<span style="color:#f92672">/</span>libBLAS.dylib
LAPACK<span style="color:#f92672">:</span> <span style="color:#f92672">/</span>Library<span style="color:#f92672">/</span>Frameworks<span style="color:#f92672">/</span>R.framework<span style="color:#f92672">/</span>Versions<span style="color:#f92672">/</span><span style="color:#ae81ff">4.0</span><span style="color:#f92672">/</span>Resources<span style="color:#f92672">/</span>lib<span style="color:#f92672">/</span>libRlapack.dylib

locale<span style="color:#f92672">:</span>
[1] zh_CN.UTF<span style="color:#ae81ff">-8</span><span style="color:#f92672">/</span>zh_CN.UTF<span style="color:#ae81ff">-8</span><span style="color:#f92672">/</span>zh_CN.UTF<span style="color:#ae81ff">-8</span><span style="color:#f92672">/</span>C<span style="color:#f92672">/</span>zh_CN.UTF<span style="color:#ae81ff">-8</span><span style="color:#f92672">/</span>zh_CN.UTF<span style="color:#ae81ff">-8</span>

attached base packages<span style="color:#f92672">:</span>
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages<span style="color:#f92672">:</span>
 [1] dplyr_1.0.2      jiebaR_0.11      jiebaRD_0.1      wordcloud2_0.2.1 ggthemes_4.2.0  
 [6] FactoMineR_2.3   factoextra_1.0.7 reshape2_1.4.4   ggpubr_0.4.0     ggplot2_3.3.2   
[11] magrittr_1.5    

loaded via a <span style="color:#a6e22e">namespace </span>(and not attached)<span style="color:#f92672">:</span>
 [1] tidyselect_1.1.0     purrr_0.3.4          haven_2.3.1          lattice_0.20<span style="color:#ae81ff">-41</span>     
 [5] carData_3.0<span style="color:#ae81ff">-4</span>        colorspace_1.4<span style="color:#ae81ff">-1</span>     vctrs_0.3.4          generics_0.0.2      
 [9] htmltools_0.5.0      yaml_2.2.1           rlang_0.4.7          pillar_1.4.6        
[13] foreign_0.8<span style="color:#ae81ff">-80</span>       glue_1.4.2           withr_2.3.0          readxl_1.3.1        
[17] lifecycle_0.2.0      plyr_1.8.6           stringr_1.4.0        munsell_0.5.0       
[21] ggsignif_0.6.0       gtable_0.3.0         cellranger_1.1.0     zip_2.1.1           
[25] htmlwidgets_1.5.2    leaps_3.1            labeling_0.3         rio_0.5.16          
[29] forcats_0.5.0        curl_4.3             broom_0.7.1          Rcpp_1.0.5          
[33] scales_1.1.1         backports_1.1.10     flashClust_1.01<span style="color:#ae81ff">-2</span>    jsonlite_1.7.1      
[37] scatterplot3d_0.3<span style="color:#ae81ff">-41</span> abind_1.4<span style="color:#ae81ff">-5</span>          farver_2.0.3         digest_0.6.25       
[41] hms_0.5.3            stringi_1.5.3        openxlsx_4.2.2       rstatix_0.6.0       
[45] ggrepel_0.8.2        grid_4.0.2           tools_4.0.2          tibble_3.0.3        
[49] cluster_2.1.0        crayon_1.3.4         car_3.0<span style="color:#ae81ff">-10</span>           tidyr_1.1.2         
[53] pkgconfig_2.0.3      ellipsis_0.3.1       MASS_7.3<span style="color:#ae81ff">-51.6</span>        data.table_1.13.0   
[57] rstudioapi_0.11      R6_2.4.1             compiler_4.0.2      
</code></pre></div>
        </p>
    </div>
    

    

    <div class="page-footer">
        
    </div>


        

<link rel="stylesheet" type="text/css" href="/css/katex.min.css" crossorigin="anonymous">
<script type="text/javascript" src="/js/katex.min.js" crossorigin="anonymous"></script>
<script type="text/javascript" src="/js/auto-render.min.js"onload="renderMathInElement(document.body);" crossorigin="anonymous"></script>

        </div>
        <footer class="footer-mobile">
	<div class="social-icons">
        
    <a class="social-icon" href="mailto:xwei97@zju.edu.cn" target="_blank" rel="noopener" title="Email">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M25.2794292,5.59128519 L14,16.8707144 L2.72057081,5.59128519 C3.06733103,5.30237414 3.51336915,5.12857603 4,5.12857603 L24,5.12857603 C24.4866308,5.12857603 24.932669,5.30237414 25.2794292,5.59128519 Z M25.9956978,6.99633695 C25.998551,7.04004843 26,7.08414302 26,7.12857603 L26,20.871424 C26,21.0798433 25.9681197,21.2808166 25.9089697,21.4697335 L18.7156355,14.2763993 L25.9956978,6.99633695 Z M24.9498374,22.6319215 C24.6672737,22.7846939 24.3437653,22.871424 24,22.871424 L4,22.871424 C3.5268522,22.871424 3.09207889,22.7071233 2.74962118,22.432463 L10.0950247,15.0870594 L13.9848068,18.9768415 L14.1878486,18.7737996 L14.2030419,18.7889929 L17.6549753,15.3370594 L24.9498374,22.6319215 Z M2.00810114,21.0526627 C2.00273908,20.9929669 2,20.9325153 2,20.871424 L2,7.12857603 C2,7.08414302 2.00144896,7.04004843 2.00430222,6.99633695 L9.03436454,14.0263993 L2.00810114,21.0526627 Z"></path>
        </svg>
    </a>
    

    

    
    <a class="social-icon" href="https://twitter.com/LexiWeiXin" target="_blank" rel="noopener" title="Twitter">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M8.991284,24.971612 C19.180436,24.971612 24.752372,16.530224 24.752372,9.210524 C24.752372,8.970656 24.747512,8.731868 24.736496,8.494376 C25.818008,7.712564 26.758256,6.737 27.5,5.62622 C26.507372,6.067076 25.439252,6.364292 24.318752,6.498212 C25.462472,5.812628 26.340512,4.727444 26.754584,3.434036 C25.684088,4.068536 24.499004,4.53002 23.23724,4.778528 C22.226468,3.701876 20.786828,3.028388 19.193828,3.028388 C16.134404,3.028388 13.653536,5.509256 13.653536,8.567492 C13.653536,9.0023 13.702244,9.424904 13.797176,9.830552 C9.19346,9.599108 5.11106,7.39472 2.3792,4.04294 C1.903028,4.861364 1.629032,5.812628 1.629032,6.827072 C1.629032,8.74904 2.606972,10.445612 4.094024,11.438132 C3.185528,11.41016 2.331788,11.160464 1.585184,10.745096 C1.583888,10.768208 1.583888,10.791428 1.583888,10.815728 C1.583888,13.49888 3.493652,15.738584 6.028088,16.246508 C5.562932,16.373084 5.07326,16.44134 4.56782,16.44134 C4.210988,16.44134 3.863876,16.406024 3.526484,16.34144 C4.231724,18.542264 6.276596,20.143796 8.701412,20.18894 C6.805148,21.674696 4.416836,22.56008 1.821488,22.56008 C1.374476,22.56008 0.93362,22.534592 0.5,22.4834 C2.951708,24.054476 5.862524,24.971612 8.991284,24.971612"></path>
        </svg>
    </a>
    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
    
    
    <a class="social-icon" href="https://github.com/lexinwei/" target="_blank" rel="noopener" title="GitHub">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M13.9988029,1.32087331 C6.82105037,1.32087331 1,7.14112562 1,14.3212723 C1,20.0649109 4.72454649,24.9370678 9.89038951,26.6560892 C10.5408085,26.7757983 10.7778323,26.374374 10.7778323,26.0296121 C10.7778323,25.7215609 10.7666595,24.9035493 10.760275,23.8189856 C7.14426471,24.6042767 6.38131925,22.0760223 6.38131925,22.0760223 C5.78995672,20.5740732 4.93762853,20.1742451 4.93762853,20.1742451 C3.75729765,19.3682044 5.02701126,19.3841656 5.02701126,19.3841656 C6.33183953,19.4759425 7.01817121,20.7241085 7.01817121,20.7241085 C8.17775254,22.7104801 10.0611744,22.1366749 10.8017741,21.8038838 C10.919887,20.9643246 11.2558703,20.3913175 11.6269683,20.066507 C8.74038491,19.7385043 5.70536235,18.6228163 5.70536235,13.6413251 C5.70536235,12.2223743 6.21213051,11.0611968 7.04370914,10.1530044 C6.90963504,9.82420367 6.46351945,8.50181809 7.17139875,6.71256734 C7.17139875,6.71256734 8.26234691,6.36301702 10.7459099,8.04532771 C11.78259,7.75642995 12.8950858,7.61277914 14.000399,7.60719272 C15.1049142,7.61277914 16.2166119,7.75642995 17.2548881,8.04532771 C19.736855,6.36301702 20.8262071,6.71256734 20.8262071,6.71256734 C21.5356825,8.50181809 21.0895669,9.82420367 20.9562909,10.1530044 C21.7894656,11.0611968 22.2922435,12.2223743 22.2922435,13.6413251 C22.2922435,18.6355852 19.2524325,19.734514 16.3570705,20.0561322 C16.8231376,20.4575564 17.2389269,21.2508282 17.2389269,22.4638795 C17.2389269,24.2012564 17.2229657,25.603448 17.2229657,26.0296121 C17.2229657,26.3775663 17.4575954,26.7821827 18.116793,26.6552912 C23.2786458,24.9322794 27,20.0633148 27,14.3212723 C27,7.14112562 21.1789496,1.32087331 13.9988029,1.32087331"></path>
        </svg>
    </a>
    

    
    <a class="social-icon" href="https://stackoverflow.com/users/11497074/x-wei" target="_blank" rel="noopener" title="Stack Overflow">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M20.8863636,23.7090909 L20.8863636,17.3818182 L22.9863636,17.3818182 L22.9863636,25.8090909 L4.03181818,25.8090909 L4.03181818,17.3818182 L6.13181818,17.3818182 L6.13181818,23.7090909 L20.8863636,23.7090909 Z M8.45,16.7818182 L8.88636364,14.7090909 L19.1954545,16.8636364 L18.7590909,18.9363636 L8.45,16.7818182 Z M9.81363636,11.8727273 L10.6863636,9.93636364 L20.2318182,14.4090909 L19.3590909,16.3181818 L9.81363636,11.8727273 Z M12.4590909,7.18181818 L13.7954545,5.57272727 L21.8954545,12.3090909 L20.5590909,13.9181818 L12.4590909,7.18181818 Z M17.6954545,2.19090909 L23.9681818,10.6454545 L22.2772727,11.9 L16.0045455,3.44545455 L17.6954545,2.19090909 Z M8.23181818,21.5818182 L8.23181818,19.4818182 L18.7590909,19.4818182 L18.7590909,21.5818182 L8.23181818,21.5818182 Z"></path>
        </svg>
    </a>
    
    
    

    

    
    <a class="social-icon" href="https://scholar.google.com/citations?user=r7VSdzkAAAAJ&amp;hl=en" target="_blank" rel="noopener" title="Google Scholar">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
             <path d="M15.7399177,5.48151786 C15.149924,4.82272859 14.4663239,4.49480724 13.6933878,4.49480724 C12.8185607,4.49480724 12.1471312,4.80966127 11.6791845,5.43796013 C11.2112379,6.06433732 10.9771365,6.82326666 10.9771365,7.71483354 C10.9771365,8.47423262 11.1052474,9.24733956 11.3616402,10.035649 C11.6170934,10.8236168 12.0350768,11.5270314 12.6210136,12.1470458 C13.2049434,12.7685974 13.8844439,13.0787968 14.6564833,13.0787968 C15.5171327,13.0787968 16.189459,12.7903763 16.671626,12.2137916 C17.1517432,11.6379755 17.3927199,10.9111166 17.3927199,10.035649 C17.3927199,9.29000051 17.2655484,8.50924961 17.0102233,7.69296927 C16.7561365,6.87562134 16.3319184,6.13812924 15.739875,5.48156056 L15.7399177,5.48151786 Z M24.3072528,4.96950105 L24.3072528,13.3462071 C24.3072528,13.7339989 23.9898792,14.0515006 23.6019592,14.0515006 L23.3422783,14.0515006 C22.9543156,14.0515006 22.6369848,13.7341271 22.6369848,13.3462071 L22.6369848,4.96950105 C22.6369848,4.62513879 22.5869788,4.33765779 23.068249,4.27714671 L23.068249,3.16262405 L19.3634081,6.20120253 C19.4061972,6.28058863 19.446851,6.33132057 19.4854124,6.39789556 C19.8110278,6.97418137 19.9768461,7.69087679 19.9768461,8.56681414 C19.9768461,9.23824368 19.8649625,9.84134739 19.638078,10.3740328 C19.4122611,10.9065473 19.1375912,11.3415267 18.817143,11.6765369 C18.496652,12.0127855 18.1752643,12.3201237 17.8538766,12.5967152 C17.5324462,12.8739047 17.2577763,13.162923 17.0328135,13.4629589 C16.8060571,13.7620125 16.6932767,14.0712297 16.6932767,14.3917206 C16.6932767,14.7126813 16.8397076,15.0387237 17.1317151,15.3673283 C17.4226551,15.6968724 17.7805972,16.0162103 18.2038758,16.3317049 C18.6279231,16.6450216 19.0510736,16.9928002 19.4743522,17.3712399 C19.8985703,17.74921 20.2545052,18.2354765 20.5454452,18.8259399 C20.8385631,19.4186239 20.9848231,20.071093 20.9848231,20.7882582 C20.9848231,21.7343577 20.7435475,22.5887297 20.2626616,23.3491963 C19.7802812,24.1064175 19.1518542,24.7106315 18.3796867,25.1537246 C17.6054695,25.599807 16.777531,25.9355858 15.8945476,26.1623422 C15.0097279,26.3870488 14.1317408,26.5 13.2548639,26.5 C12.7014246,26.5 12.1429889,26.4573818 11.5814359,26.3697112 C11.0178757,26.2822114 10.4532907,26.1276668 9.8847343,25.9098782 C9.31502485,25.6906377 8.81052389,25.4214338 8.37319577,25.0981672 C7.93475736,24.778146 7.58181166,24.3652016 7.3111559,23.862238 C7.04054285,23.3591036 6.90615445,22.7936645 6.90615445,22.1655364 C6.90615445,21.4203577 7.1136515,20.7291563 7.52975591,20.0847582 C7.94586031,19.4451002 8.49707907,18.9108775 9.18272893,18.487727 C10.3789864,17.7434877 12.2558547,17.2836547 14.8102164,17.1098936 C14.2262867,16.3799173 13.9333396,15.6928582 13.9333396,15.0498693 C13.9333396,14.6841125 14.0289531,14.2920503 14.218173,13.8687717 C13.9130554,13.9113899 13.5987992,13.9355175 13.2781801,13.9355175 C11.9059836,13.9355175 10.7472627,13.4894351 9.80628769,12.5911638 C8.86535538,11.6944724 8.39535897,10.5719642 8.39535897,9.21488479 C8.39535897,9.07298054 8.39941581,8.94802965 8.40970739,8.8095417 L2.83316537,8.8095417 L11.2417283,1.5 L25.1668346,1.5 L23.87996,2.50451805 L23.87996,4.27748834 C24.3577712,4.3388962 24.3072955,4.62607827 24.3072955,4.96950105 L24.3072528,4.96950105 Z M15.8172967,17.9839947 C15.6566455,17.9549562 15.430914,17.9386434 15.1399313,17.9386434 C14.512273,17.9386434 13.8947782,17.9941154 13.2885571,18.1048887 C12.6822934,18.2126727 12.084186,18.3947184 11.4941496,18.6515809 C10.9019353,18.906906 10.423996,19.2823566 10.0597766,19.7787865 C9.69355004,20.274576 9.51150436,20.858463 9.51150436,21.5303623 C9.51150436,22.1706609 9.67215551,22.7419077 9.99371402,23.2394479 C10.3142049,23.7338281 10.7373127,24.1208513 11.2632509,24.3980408 C11.7891891,24.6756999 12.3405787,24.8854176 12.916224,25.0231795 C13.4929795,25.1600875 14.0799412,25.2311463 14.6780913,25.2311463 C15.8610679,25.2311463 16.8794647,24.9647609 17.7338794,24.4320756 C18.5863724,23.899561 19.0135798,23.0772595 19.0135798,21.9672634 C19.0135798,21.7338025 18.9810823,21.5040141 18.916856,21.2796064 C18.8498112,21.0527219 18.7835352,20.8585057 18.7185829,20.6972567 C18.6534171,20.5389543 18.529363,20.3483679 18.3462498,20.1286576 C18.1642041,19.9083067 18.0248194,19.7455631 17.9310848,19.6372239 C17.8354714,19.5253403 17.6565431,19.3656714 17.3929761,19.1545445 C17.131587,18.9425209 16.9647012,18.8103104 16.8904822,18.7614574 C16.8172027,18.7100422 16.6270434,18.5711699 16.3207728,18.3458227 C16.0147157,18.1186393 15.8468903,17.9981723 15.8173394,17.983952 L15.8172967,17.9839947 Z"></path>
        </svg>
    </a>
    

    

    
    <a class="social-icon" href="https://orcid.org/0000-0001-6635-5531" target="_blank" rel="noopener" title="Orcid">
        <svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" fill="#ABABAB" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink">
            <path d="M27,14 C27,21.1804687 21.1804687,27 14,27 C6.81953125,27 1,21.1804687 1,14 C1,6.81953125 6.81953125,1 14,1 C21.1804687,1 27,6.81953125 27,14 Z M9.76484375,19.9109375 L9.76484375,13.9492188 L9.76484375,9.03359375 L8.20078125,9.03359375 L8.20078125,19.9109375 L9.76484375,19.9109375 Z M12.0601563,9.03359375 L12.0601563,19.9210938 L16.3054688,19.9210938 C19.890625,19.9210938 22.0742188,17.2703125 22.0742188,14.4773437 C22.0742188,11.9078125 20.3070313,9.03359375 16.2851563,9.03359375 L12.0601563,9.03359375 Z M13.6242188,18.509375 L13.6242188,10.4453125 L16.03125,10.4453125 C19.078125,10.4453125 20.4695313,12.29375 20.4695313,14.4773437 C20.4695313,15.8179688 19.6570313,18.509375 16.1125,18.509375 L13.6242188,18.509375 Z M10.0085938,6.76875 C10.0085938,6.21015625 9.5515625,5.74296875 8.9828125,5.74296875 C8.4140625,5.74296875 7.95703125,6.2 7.95703125,6.76875 C7.95703125,7.32734375 8.4140625,7.79453125 8.9828125,7.79453125 C9.5515625,7.79453125 10.0085938,7.32734375 10.0085938,6.76875 Z"></path>
        </svg>
    </a>
    

    

    

    

</div>




	<div class="footer-mobile-links">
        <p><a href="https://github.com/kimcc/hugo-theme-noteworthy" target="_blank" rel="noopener">Noteworthy theme</a></p>
		<span class="divider-bar">|</span>
        <p><a href="https://gohugo.io" target="_blank" rel="noopener">Built with Hugo</a></p>
	</div>

	<script src="https://lexinwei.github.io/js/main.min.fa5c2b23e07b5d9bfad267a52b4b24fdb053e6fb7524993383594926a3ac270c.js" integrity="sha256-+lwrI+B7XZv60melK0sk/bBT5vt1JJkzg1lJJqOsJww=" crossorigin="anonymous"></script>
</footer>

    </body>
</html>